### ISSUE TITLE
Trainer reports -inf co-metrics, and co-metrics need not match the saved checkpoint

### ISSUE THREAD

**Issue by magilliam27** [magilliam27/MCI-GRU#111]
**Current ownership and scope (2026-09-19):** Owner-requested GPT-6 Astra / Extra High task reserved for `magilliam27`. No native blockers. Begin with the concrete public `Trainer.train` / `TrainingResult` / saved-checkpoint boundary proposal; it is not yet owner-confirmed. Prepare the coordinated consumer plan before changing optional-value behavior. Do not leave ensemble logging or saved summaries incompatible.

**Owned paths:** `mci_gru/training/trainer.py` (selected-checkpoint co-metric storage/result only) and new `tests/test_checkpoint_metrics.py`. Nothing else initially. #144 retains later execution-event reporting in trainer/ensemble; shared summary consumers and registry are coordinated extensions before edits.

## Summary

`TrainingResult.best_val_ic` and `TrainingResult.best_val_rank_ic` can be reported as `-inf`, and even when finite they do not necessarily belong to the epoch whose checkpoint was saved.

Found while rebuilding `docs/ARCHITECTURE.md` from current code (part of #101). No code change was made; this issue records the behavior.

## Mechanism

Both co-metrics initialize to negative infinity:

```python
self.best_val_ic = float("-inf")
self.best_val_rank_ic = float("-inf")
```

and are only assigned inside the `improved` branch, guarded on availability:

```python
            if improved:
                self.best_val_loss = val_loss
                if validation.ic is not None:
                    self.best_val_ic = validation.ic
                if validation.rank_ic is not None:
                    self.best_val_rank_ic = validation.rank_ic
```

`best_val_loss` is unconditional, but the two IC metrics are not. They are then returned verbatim:

```python
        return TrainingResult(
            best_val_loss=self.best_val_loss,
            best_val_ic=self.best_val_ic,
            best_val_rank_ic=self.best_val_rank_ic,
```

## Two distinct failure modes

**1. Sentinel leaks into reported evidence.**

`ValidationObservation.ic` is `None` whenever no validation rows are eligible. The fail-closed guard added in #104 only raises for the *selected* metric, and `val_loss` has no eligible-row condition, so it never raises. A run with `selection_metric=val_loss` and zero eligible IC rows therefore trains to completion and reports `best_val_ic = -inf`.

That value reaches `training_summary.json`, and `merge_walkforward_summary` filters only on `None`:

```python
    ics = [s["mean_best_val_ic"] for s in summaries if s.get("mean_best_val_ic") is not None]
```

`-inf` passes that filter, so a single affected window drives `mean_best_val_ic_across_windows` to `-inf` for the entire walk-forward run.

**2. Co-metric provenance does not match the checkpoint.**

Even with finite values, a co-metric is only refreshed on epochs where it happened to be available *and* the run improved. If epoch 4 improves with IC available and epoch 9 improves with IC unavailable, the saved checkpoint is epoch 9's while the reported `best_val_ic` is epoch 4's. Nothing in `TrainingResult` distinguishes the two, so the reported figure silently describes a different model than the one on disk.

This is why `docs/ARCHITECTURE.md` now hedges rather than claiming co-metrics come from the selected epoch.

## Suggested direction (not prescriptive)

Options worth weighing:

- Make the co-metrics `float | None` and emit `null` rather than a sentinel, consistent with `ValidationObservation` already using `None` for "no eligible rows".
- Record them only when they belong to the selected epoch, and otherwise report `None`.
- Alternatively harden the consumer so `merge_walkforward_summary` rejects non-finite values instead of averaging them.

The first two address the cause; the third only limits the blast radius.

## Reproduction sketch

Run with `training.selection_metric=val_loss` against a validation split whose labels are entirely NaN, so no IC rows are eligible. Training completes and `training_summary.json` carries `-inf` for both IC metrics.

## References

- `mci_gru/training/trainer.py` (initializers, `improved` branch, `TrainingResult` construction)
- `mci_gru/walkforward.py` (`merge_walkforward_summary`)
- #104 introduced the fail-closed selection guard, which covers the selected metric only
- Related in spirit to #107: a degenerate configuration producing green-but-meaningless numbers

**Comment by magilliam27** [comment]
## This is unfinished scope from #104, not a new discovery

Re-reading the #98 ownership manifest while planning the remaining slices, the `codex/validation-selection-coverage` entry — the slice that shipped as #104 — states its intent as:

> Intent: minimum eligible validation-row coverage and fail-closed IC/rank-IC checkpoint selection. **Gate val-loss runs with no eligible IC rows so unavailable co-metrics are explicit rather than initial `-inf` sentinels.** Verify `N-1/N`, both IC metrics, val-loss/no-IC, **same-checkpoint co-metrics**, focused trainer tests, Ruff, and `git diff --check`.

Both failure modes described in this issue were therefore anticipated before extraction began, and both were listed as required verification:

- "unavailable co-metrics are explicit rather than initial `-inf` sentinels" is the sentinel-leak mode.
- "same-checkpoint co-metrics" is the provenance mode, where a finite co-metric belongs to an earlier improved epoch than the checkpoint actually saved.

#104 implemented only the first half of the intent — the `minimum_selection_rows` coverage guard and fail-closed selection — and left the co-metric half unaddressed. The `-inf` initializers and the conditional updates inside the `improved` branch are unchanged from before that PR.

This does not change the technical content of this issue, but it does change how it should be prioritised and framed:

1. It is **completing a merged slice**, not new work, so it inherits that slice's already-agreed intent rather than needing fresh justification.
2. The manifest's verification list is effectively a ready-made acceptance criteria set: `N-1/N` boundary, both IC metrics, the val-loss-with-no-IC case, and same-checkpoint co-metrics.
3. It suggests re-reading each merged slice's manifest intent against what actually shipped, in case other slices were also landed partially. That check has not been done.

Filed for the record so the scope history is not lost.

**Comment by magilliam27** [comment]
## Partial mitigation in #130 — labelling only, mechanism untouched

#130 relabels the ensemble completion log from

```
Model N training complete. Best val loss: …, best val IC: …, best val Rank IC: …
```

to

```
Model N training complete. Selected-checkpoint val loss: …, val IC: …, val Rank IC: …
```

The word "best" asserted precisely the property this issue records as not holding: that the reported co-metrics belong to the epoch whose checkpoint was saved. The new wording claims only that the loss belongs to the selected checkpoint and stops short of claiming the IC metrics do.

**This does not fix either failure mode.** `best_val_ic` and `best_val_rank_ic` still initialise to `-inf` and are still assigned only inside the guarded `improved` branch, so they can still be reported as `-inf`, and when finite they can still belong to a different epoch than the saved checkpoint. This issue stays open.

The change was salvaged from an unregistered worktree recovered during the #99 topology audit, where it existed as an uncommitted modification. Provenance and the full salvage rationale are on #130.

**Comment by magilliam27** [comment]
## Cross-link: same shape as #144

This ticket and #144 (`Capture environment and code identity in training run artifacts`, a child of the #140 map) are the same defect shape stated at two layers: **the artifact does not describe the run that produced it.**

- Here: `TrainingResult.best_val_ic` and `best_val_rank_ic` can be `-inf`, and even when finite they need not belong to the epoch whose checkpoint was saved. The reported co-metric does not describe the saved artifact.
- #144: a run's artifacts do not record the environment or code identity that produced them.

Whoever picks up #144 should read this one first — a provenance record that faithfully captures the environment while still reporting co-metrics that belong to a different epoch would leave the same question unanswerable, just with more metadata around it.

**Deliberately a cross-link, not a parent/child edge.** #144 is a `wayfinder:task` leaf under #140; hanging a child off it would nest one level deeper than the map convention uses, and would pull this ticket into a map whose declared scope is the execution substrate rather than trainer accounting. If you would rather it were absorbed into #140 alongside #132, say so and I will wire it — it is a one-call change.

#144 remains open and is unaffected by this note.

**Comment by magilliam27** [comment]
Owner requested dedicated GPT-6 Astra / Extra High sessions to advance the project. This ticket is now claimed with the bounded initial paths declared in its body before branching. Owner-requested GPT-6 Astra / Extra High task reserved for `magilliam27`. No native blockers. Begin with the concrete public `Trainer.train` / `TrainingResult` / saved-checkpoint boundary proposal; it is not yet owner-confirmed. Prepare the coordinated consumer plan before changing optional-value behavior. Do not leave ensemble logging or saved summaries incompatible. Shared-file extensions require coordination; no merge, GPU/vendor work or shared-environment changes are included.
